### ISSUE TITLE
Using Raw string literal breaks GLSL highlighting.

### ISSUE THREAD

**Issue by Raivr-dev** [danielscherzer/GLSL#35]
### Installed product versions
- Visual Studio 2017 Community
- This extension: 0.5.67

### Description
I want to include my shader code as follows:

const GLchar* sFragmentShaderCode_ = 
#include "myShader.frag"

That works fine when I use a Raw string literal, which makes the first line of myShader.frag this:
R"(
Unfortunately, GLSL Language Integration doesn't show highlighting because of that line.

### Steps to recreate
See above.

### Current behavior
See above.
Maybe it's not wrong, but how can I get highlighting when the first line of the shader code is 
R"(
?

### Expected behavior
Highlight my shader code even if the first line is R"(

**Comment by danielscherzer** [comment]
I'm not quite sure I understand the problem statement: 
`const GLchar* sFragmentShaderCode_ = #include "myShader.frag"`
seems to mee to be C++ code. If I understand correctly you use the `#include` statement to put the shader code inside a string. This only works for raw string literals. So you put `R"(` at the start of your shader file (`myShader.frag`) and syntax highlighting of this shader file stops to work?
I quickly tried this:

![screenshot](https://user-images.githubusercontent.com/8530595/62120732-7225f300-b2c2-11e9-8856-cb500a3ad836.PNG)

and syntax highlighting still works as expected...

**Comment by RaymondSmit** [comment]
Yes, you understood correctly.
See attached pictures. When I comment out the `R"(` or don't use it at all, there is syntax highlighting. But with the `R"(` at the start of the shader file, syntax highlighting stops working.


![image](https://user-images.githubusercontent.com/10847452/62150265-d023fc00-b2fd-11e9-8ba4-b8f0af1ae66b.png)

![image](https://user-images.githubusercontent.com/10847452/62150305-f184e800-b2fd-11e9-9161-d3a9765aed12.png)

**Comment by danielscherzer** [comment]
I tried exactly the same code as you did:

![Unbenannt](https://user-images.githubusercontent.com/8530595/62191651-3481a280-b374-11e9-8b92-711e60a11277.PNG)

and again it works as expected. My only guess is that maybe the `R"(` has some nasty invisible code that screws up parsing. Could you try and remove `R` then `"` ... to see if the problem persists, or you send me your source file and I try, if I get the missing highlighting.

**Comment by RaymondSmit** [comment]
It's definitely the 'R' followed by the double quote that screws it up.
Here's a test file. Remove the ".txt" extension (I couldn't upload it otherwise).
Even if I open just this file in Visual Studio, with no solution loaded, the highlighting stops working. Remove the `R"(` , scroll the code out of view and back in view, then highlighting works.

[testshader.frag.txt](https://github.com/danielscherzer/GLSL/files/3452214/testshader.frag.txt)

**Comment by danielscherzer** [comment]
Thanks! Now it makes sense: I use the Visual Studio built in Cpp classifier for finding identifiers, comments, ... Then I look at each identifiers and use a dictionary lookup into a list of predefined glsl variables, keywords and functions to add them as additional classes, so I do not have to implement a parser for a c style language, like glsl.

As soon as you put in `R"(` the cpp classifier, classifies everything following as a raw string literal, so no identifiers for me to look up...

I think changing this behaviour would either mean not using the cpp classifier at all or putting in a hack for `R"(`. As this is a "special" use case I would prefer to go another way:

If I understand your usage correctly you want a compile time string generated from a shader file.
The `R"(` is a hack, so you can use the `#include` mechanism for doing so. This could for instance break if the implementation of the `#include` mechanism changes.
I used [T4 design time text templates](https://docs.microsoft.com/en-us/visualstudio/modeling/code-generation-and-t4-text-templates?view=vs-2019) for generating code at compile time. Could those be a viable solution to your problem too?

**Comment by RaymondSmit** [comment]
I looked at T4. Thanks for the suggestion, but it seems too complex in my case. If I understand correctly, with T4 I would be generating my .cpp (or .h) file. Maybe that is just moving the problem somewhere else.
I tried to get the `R"(` into a macro, but I didn't succeed. A single `"` seems to be illegal in macro definition.
How do you ship your glsl source code? In separate files next to the .exe?

**Comment by danielscherzer** [comment]
For shipping glsl code I use mainly resources embedded into the executable. I work mainly with C#. Here it is as easy as marking a project file as embedded resource in its' properties, then it is automatically incldued into the exe/dll file during building and you can access it as a stream during execution.

**Comment by RaymondSmit** [comment]
All right, I'll use resource files. Thanks for your help. And thanks for making the GLSL syntax highlighter!

**Comment by danielscherzer** [comment]
I did my own parser in the new version of t he extension, so this hopefully removes this issue.

**Comment by RaymondSmit** [comment]
Great!
I'll give it a try.

**Comment by Raivr-dev** [comment]
Works perfectly now!

**Comment by roquo** [comment]
Downloaded the current version from [CI Build](http://vsixgallery.com/extension/b62242eb-0ae5-4494-b013-6158ade63816/) and have some issues. Didn't wanted to open a new issue, maybe this is a desired behavior.
Working with Visual Studio Professional 2017.

I get different false warning and errors depending on where the first line is:
![image](https://user-images.githubusercontent.com/35525625/83106862-7cb94080-a0bd-11ea-8c90-f142c9dcceda.png)

![image](https://user-images.githubusercontent.com/35525625/83106893-8a6ec600-a0bd-11ea-9e2a-39abf76eb953.png)

Highlighting works fine, but the parser is complaining about false errors and warnings, and doesn't recognize real errors, example:
![image](https://user-images.githubusercontent.com/35525625/83106969-ad00df00-a0bd-11ea-8f31-6c0bfa867c35.png)

The errors I get:
![image](https://user-images.githubusercontent.com/35525625/83107024-c86bea00-a0bd-11ea-9f08-dc187648bd9e.png)

Is there something we can do about this? Because I really want and have to use raw string literals

**Comment by danielscherzer** [comment]
To clarify what my extension does: I only parse the shader text for syntax coloring. 
The errors and warnings are generated by whatever shader compiler the extension is using.
And then I parse this output to generate error and warning messages and squiggly lines.
This shader compiler could be inside the driver of your graphics card or any exe specified as an option to the extension. The default is to use the graphcis cards driver.

Since the shader text is handled over to this shader compiler, it will probably recognize `R"(` as beeing wrong in the context of a shader. Dependent on the compiler it will then either stop checking errors (my nvidia graphics card driver does this) or put out some other more or less reasonable messages. In your case it probably ignored the version string and so assumed an old glsl version.

As I understand it, putting `R"(` is a hack to quickly get a shader string from a file. You could also do this using for instance resources our some form of compile time code generation. This I also discussed with the initiator of this issue and he did afterwards switch to the use of resources.

Because this of this hackish nature, I want to avoid removing the string `R"(` from the beginning of every shader that I pass on to a shader compiler. But if you have suggestion for a clean way of supporting this kind of behaviour in the extension, I'm happy to include it.

Maybe an option to remove a given preamble from a shader file woudl do the trick. But I would be interested why you have to use raw string literals as you stated above?

**Comment by roquo** [comment]
Thanks for your quick reply! I putted it now into the resources. This works fine, too. I only have to do some marshaling, but that should be ok! I really like your tool and I also presented it my team already, they really like the syntax highlighting features and are now willing to implement more features in shaders, because of a faster feedback on errors etc.!
Before your tool was provided to the developers, we used to implement shaders directly in the c++ code like this 
![image](https://user-images.githubusercontent.com/35525625/83115108-30283200-a0ca-11ea-96ce-cf0c77cb459c.png)

Your tool is awseome, thanks!!
